### PR TITLE
Metrics update

### PR DESCRIPTION
**Pull request by ArneRobben**
## What does this change?

This PR adds in a few things:

- I renamed `preprocess.py` to `train_test_split.py`. It felt to me that most of the preprocessing is now done in src/data_processing/* and the main utility of the old `preprocess.py` was just to split the data in train and test
- a separate `metrics.py` file where all metrics are calculated
- we now calculate the following metrics:
   - (as before) the usual 'lowest level' HRCS RAC tags, metrics both on an aggregate (such as F1 macro & Micro) and on a per label level
   - we also calculate the parent level metrics, where we take the first digit of a lowest level classification to infer the parent. E.g. `"1.1: Normal biological development and functioning (underpinning)"` maps to `"1: Underpinning research"`
   - to get a sense of how the metrics do on Wellcome's grants, we also calculate metrics when test data is filtered for Wellcome grants only. In order to do this, a train_meta.parquet and test_meta.parquet file are created which contains the FundingOrganisation field so we can filter for Wellcome. The metrics basically filter the test data for grants that are only related to Wellcome. This is likely to perform a little worse because of underrepresentation of some of the classes but Wellcome is still the second largest funding organisation in the data. We also report on the parent level and super parent level of Wellcome only metrics.

## How to test

`make train_test_split` and `make train_ra`

## How can we measure success?

Success in this case if the code does what it needs to do and if the metrics we care about (listed above) are logged in wandb

## Have we considered potential risks?

The classifier might be considered worse for "Wellcome only" in the test set. The onus of explaining this is on us

### CODE REVIEW

File: src/train_test_split.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+          assumes 'text' is the last column and all prior columns are labels.
+    """
+    df = df.copy()
+    df.dropna(subset=category, inplace=True)
+
+    # Randomly shuffle dataframe and reset index; add a stable row id
+    df = df.sample(frac=1, random_state=10).reset_index(drop=True)
+    df["_row_id"] = np.arange(len(df), dtype=np.int64)
+
+    # Labels (multilabel binarized)
```
Comment (justinbt1) [inline src/train_test_split.py]: I think this may be obvious from below Class and variable names, consider removing this comment?
Think of comments as simply another line of code we have to maintain.
See https://ds.wellcome.data/python_standards/#comments

File: src/train.py [hunk 2]
```diff
@@ -221,107 +187,32 @@ def train(train_data, test_data, model_path, config, class_counts, class_weighti
...
-    print(df)
-
-    table = wandb.Table(
-        data=[list(values) for values in data],
-        columns=["label", "precision", "recall", "f1", "train_count", "test_count"],
-    )
-
-    wandb.log({"metrics and value_count table": table})
+    print(f"Model initialized using {model_class.__name__}")
+    return model
```
Comment (justinbt1) [inline src/train.py]: Good idea to split this out into a seperate function.

File: src/train.py [hunk 3]
```diff
@@ -221,107 +187,32 @@ def train(train_data, test_data, model_path, config, class_counts, class_weighti
...
-    )
-    print(df)
-
-    table = wandb.Table(
-        data=[list(values) for values in data],
-        columns=["label", "precision", "recall", "f1", "train_count", "test_count"],
-    )
-
-    wandb.log({"metrics and value_count table": table})
+    print(f"Model initialized using {model_class.__name__}")
```
Comment (justinbt1) [inline src/train.py]: Nice use of the `__name__` dunder method!

**Review by justinbt1 — COMMENTED** [review]
Quick review as I know you are still working on this, the data processing and metrics appear to produce the intended results.

However there are areas where the code deviates from our standards and may require refactoring due to complexity/readability. See the comments below for more information, happy to catch up in the office to chat about this.

**Review by justinbt1 — CHANGES_REQUESTED** [review]

File: src/metrics.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        dict: Dictionary mapping label IDs to label names.
+    """
+    if label_type == "long":
+        filepath = os.path.join(label_names_dir, "label_mapping_long.json")
+    elif label_type == "short":
+        filepath = os.path.join(label_names_dir, "label_mapping_short.json")
+    elif label_type == "tiny":
+        filepath = os.path.join(label_names_dir, "label_mapping_tiny.json")
+    else:
+        raise ValueError("label_type must be 'long', 'short', or 'tiny'")
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py:27]: yes but the proposed way points the file to the correct labelmapping in the if and elif statements

File: src/metrics.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                    continue  # Skip if no valid parent category
+            else:
+                continue  # Skip if no valid parent category
+
+            if predictions[sample_idx, label_idx] == 1:
+                parent_predictions[sample_idx, idx] = 1
+            if labels[sample_idx, label_idx] == 1:
+                parent_labels[sample_idx, idx] = 1
+    
+    return parent_predictions, parent_labels
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py:93]: Given predictions (np.array of length 48 of 1s and 0s containing outputs of the model), labels (np.array of lenght 48 of the true labels), and the detailed label names, this function returns parent_predictions and parent_labels, which is basically just a roll up of the predictions and labels to either the parent (8 categories) or super parent (3 categories). So the outputs can then be used to calculate metrics for parents and super parents.

File: src/metrics.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+        f"{prefix}precision/{label_names[idx]}": float(precision[idx])
+        for idx in range(indexed_len)
+    } | {
+        # Recall per class
+        f"{prefix}recall/{label_names[idx]}": float(recall[idx])
+        for idx in range(indexed_len)
+    }
+
+    table = wandb.Table(
+        columns=["prefix", "precision_macro", "precision_micro", "recall_macro", "recall_micro", "f1_macro", "f1_micro", "train_count", "test_count"],
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py]: I ran metrics.py through black

File: src/train_test_split.py (hunk 1 above)
Comment (ArneRobben) [reply src/train_test_split.py]: sure will remove it

File: src/train_test_split.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        - train_df/test_df keep original schema: label columns + final 'text' column.
+        - metadata is saved separately to avoid breaking training code that
+          assumes 'text' is the last column and all prior columns are labels.
+    """
+    df = df.copy()
+    df.dropna(subset=category, inplace=True)
+
+    # Randomly shuffle dataframe and reset index; add a stable row id
+    df = df.sample(frac=1, random_state=10).reset_index(drop=True)
+    df["_row_id"] = np.arange(len(df), dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py]: sure will use the dataset index instead

File: src/train_test_split.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    # Compose label+text DataFrames (text must remain the last column)
+    train = pd.DataFrame(y_train, columns=mlb.classes_)
+    test = pd.DataFrame(y_test, columns=mlb.classes_)
+    train["text"] = text_train
+    test["text"] = text_test
+
+    # Prepare sidecar metadata DataFrames aligned row-by-row to train/test
+    metadata_cols = [c for c in [
+        "FundingOrganisation",
+        "OrganisationReference"
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py]: Yeah that's right. We only need FundingOrganisation to filter for Wellcome later. Might leave it in though as it doesn't harm the meta file

File: src/train_test_split.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    ] if c in df.columns]
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py]: sure. maybe better for this to break if the OpenAlex format changes in the future

File: src/train_test_split.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+    if metadata_cols:
+        df_meta_indexed = df.set_index("_row_id")
+        train_meta = df_meta_indexed.loc[idx_train, metadata_cols].reset_index(drop=True)
+        test_meta = df_meta_indexed.loc[idx_test, metadata_cols].reset_index(drop=True)
+    else:
+        train_meta = pd.DataFrame(index=np.arange(len(train)))
+        test_meta = pd.DataFrame(index=np.arange(len(test)))
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py]: let's simplify and assume metadata_cols is there

File: src/train_test_split.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    return train, test, train_meta, test_meta
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py:72]: it does make sense to save down the indices for train and testing (so the train_meta and test_meta) as part of the split_data_frame function. Is there anything in particular you'd suggest pulling out to a separate function?

File: src/train_test_split.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        output_dir (str): Output directory.
+    """
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+    train.to_parquet(output_dir + "/train.parquet")
+    test.to_parquet(output_dir + "/test.parquet")
+    if train_meta is not None:
+        train_meta.to_parquet(output_dir + "/train_meta.parquet")
+    if test_meta is not None:
+        test_meta.to_parquet(output_dir + "/test_meta.parquet")
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/train_test_split.py]: good spot

File: src/metrics.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        data=[
+            [label_names[idx], 
+             float(f1_per_class[idx]), 
+             float(precision[idx]), 
+             float(recall[idx])]
+            for idx in range(indexed_len)
+        ]
+    )
+    wandb.log({f"{prefix}per_class_metrics": table_per_class})
+    wandb.log({f"{prefix}metrics and value_count table": table})
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py]: done

File: src/metrics.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+            super_parent_predictions, super_parent_labels = get_parent_categories(
+                predictions, labels, detailed_label_names, parent_level="super_parent"
+            )
+            super_parent_metrics = calculate_metrics(
+                super_parent_labels, super_parent_predictions, prefix="super_parent_"
+            )
+            all_metrics = {**detailed_metrics, **parent_metrics, **super_parent_metrics}
+        except Exception as e:
+            print(f"Warning: Could not calculate parent metrics: {e}")
+            all_metrics = detailed_metrics
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py:380]: true. My reasoning here was that after hours of training, metrics not reporting because of an issue with the parent metrics would be a shame. There are realistically not that many exceptions that can happen here so incorporated here

File: src/metrics.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+                    print("Info: test_meta.parquet not found. Run train_test_split.py to generate metadata.")
+                    wandb.log({"wellcome_metrics_status": "meta_not_found"})
+                else:
+                    meta_df = pd.read_parquet(test_meta_path)
+                    if "FundingOrganisation" not in meta_df.columns:
+                        print("Info: FundingOrganisation column missing in meta; skipping Wellcome metrics.")
+                        wandb.log({"wellcome_metrics_status": "funding_org_missing"})
+                    elif len(meta_df) != labels.shape[0]:
+                        print("Info: test_meta length does not match eval set; skipping Wellcome metrics.")
+                        wandb.log({"wellcome_metrics_status": "length_mismatch"})
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py]: do you mean why so many metrics?

The code looks at parents (8 tags) and super parents (3 tags) and then all of those for whether we filter for just Wellcome or not.

[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py]: refactored now

File: src/metrics.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+            # Loop through each sample's logits
+            for i, logit in enumerate(logits):
+                # Get the indices of the logits sorted by value in descending order
+                sorted_indices = np.argsort(logit)[::-1]
+
+                # Assign 1 to the top logits that exceed their respective thresholds
+                for rank, idx in enumerate(sorted_indices):
+                    if logit[idx] > thresholds[rank]:
+                        predictions[i, idx] = 1
+                num_tags_predicted.append(np.sum(predictions[i]))
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/metrics.py]: done

File: config/train_config.yaml [hunk 17]
```diff
@@ -3,19 +3,20 @@ preprocess_settings:
   test_train_split: 0.2
 
 training_settings:
-  category: 'RA'  # choose from "RA", "RA_top" or "HC"
+  # category: 'RA'  # choose from "RA", "RA_top" or "HC"
```
[the remark this answers is not in the document]
Comment (justinbt1) [reply config/train_config.yaml]: Commented statement still exists and may be confusing to other developers, including if this should this be commented out or not. Could this not be removed?

Commented sections like this, without a compelling reason violate our coding standards.

File: Makefile [hunk 18]
```diff
@@ -18,9 +18,9 @@ download_data:
...
 
-.PHONY: preprocess
-preprocess:
-	python src/preprocess.py \
+.PHONY: train_test_split	
+train_test_split:
+	python src/train_test_split.py \
 		"config/train_config.yaml" \
 		"data/clean/clean.parquet" \
 		"data/preprocessed"
```
[the remark this answers is not in the document]
Comment (justinbt1) [reply Makefile:26]: This will need to be updated in the README as this make target is still called `preprocess` in the newly updated README on `main`.

File: src/metrics.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+                            wandb.log({"wellcome_metrics_status": "computed"})
+                        else:
+                            print("Info: No rows for FundingOrganisation == 'Wellcome Trust' in eval set")
+                            wandb.log({"wellcome_metrics_status": "no_wellcome_rows"})
+        except Exception as e:
+            print(f"Warning: Could not compute Wellcome Trust filtered metrics: {e}")
+        
+        return all_metrics
+
+    return compute_metrics
```
[the remark this answers is not in the document]
Comment (justinbt1) [reply src/metrics.py:394]: This function is still very long and complex, any possibility of a refactor?

File: src/train.py [hunk 20]
```diff
@@ -212,7 +176,9 @@ def train(train_data, test_data, model_path, config, class_counts, class_weighti
         logging_strategy=config["training_settings"]["logging_strategy"],
     )
 
-    compute_metrics = prepare_compute_metrics(config)
+    # Provide meta_path (directory holding train_meta/test_meta) to metrics
+    meta_path = os.path.dirname(args.test_path)
```
Comment (justinbt1) [inline src/train.py]: Note: `args.test_path` is only available if train module is run as a script, as it relies on ArgParse in the `if __name__ == "__main__"` block. This means `meta_path` will not work if called from another module, notebook or test harness.

**Review by justinbt1 — COMMENTED** [review]
Thanks Arne, almost there I think, just the README update and maybe a bit more of a refactor?

File: src/train.py (hunk 20 above)
Comment (ArneRobben) [reply src/train.py]: good spot. It's now added as an explicit argument to train

**Review by justinbt1 — APPROVED** [review]
Looks good to me, LGTM!
